lighthouse-cli/test/smokehouse/smokehouse.js

```diff
@@ -1,154 +1,236 @@
   }
+  console.log(''); // extra line break
 
   return {
-    lhr,
-    artifacts,
+    id,
+    success: failingTestCount === 0,
   };
 }
 
-const cli = yargs
-  .help('help')
-  .describe({
-    'smoke-id': 'The id of the smoke test to run',
-    'debug': 'Save the artifacts along with the output',
-  })
-  .require('smoke-id', true)
-  .argv;
-
-const smokeId = cli['smoke-id'];
-const smokeTest = require('./smoke-test-dfns.js').find(smoke => smoke.id === smokeId);
-
-if (!smokeTest) {
-  process.exit(1);
-  throw new Error(`could not find smoke ${smokeId}`);
+/** @param {string} str */
+function purpleify(str) {
+  return `${log.purple}${str}${log.reset}`;
 }
 
-const lhRootDir = `${__dirname}/../../..`;
-const tmpDir = `${lhRootDir}/.tmp`;
-fs.mkdirSync(tmpDir, {recursive: true});
-const configPath = `${tmpDir}/smoke-config-${smokeTest.id}.json`;
-fs.writeFileSync(configPath, JSON.stringify(smokeTest.config));
-
-// Loop sequentially over expectations, comparing against Lighthouse run, and
-// reporting result.
-let passingCount = 0;
-let failingCount = 0;
-smokeTest.expectations.forEach(expected => {
-  console.log(`Doing a run of '${expected.lhr.requestedUrl}'...`);
-  const results = runLighthouse(expected.lhr.requestedUrl, configPath, cli.debug);
-
-  console.log(`Asserting expected results match those found. (${expected.lhr.requestedUrl})`);
-  const collated = collateResults(results, expected);
-  const counts = report(collated);
-  passingCount += counts.passed;
-  failingCount += counts.failed;
-});
-
-if (passingCount) {
-  console.log(log.greenify(`${passingCount} passing`));
+/**
+ * Run Lighthouse in the selected runner. Returns `log`` for logging once
+ * all tests in a defn are complete.
+ * @param {{requestedUrl: string, configJson?: LH.Config.Json, expectation: Smokehouse.ExpectedRunnerResult, lighthouseRunner: LighthouseRunner, retries: number, isDebug?: boolean}} testOptions
+ * @return {Promise<{passed: number, failed: number, log: string}>}
+ */
+async function runSmokeTest(testOptions) {
+  // Use a buffered LocalConsole to keep logged output so it's not interleaved
+  // with other currently running tests.
+  const localConsole = new LocalConsole();
+  const {requestedUrl, configJson, expectation, lighthouseRunner, retries, isDebug} = testOptions;
+
+  // Rerun test until there's a passing result or retries are exhausted to prevent flakes.
+  let result;
+  let report;
+  for (let i = 0; i <= retries; i++) {
+    if (i === 0) {
+      localConsole.log(`Doing a run of '${requestedUrl}'...`);
+    } else {
+      localConsole.log(`Retrying run (${i} out of ${retries} retries)...`);
+    }
+
+    // Run Lighthouse.
+    try {
+      result = await lighthouseRunner(requestedUrl, configJson, {isDebug});
+    } catch (e) {
+      logChildProcessError(localConsole, e);
+      continue; // Retry, if possible.
+    }
+
+    // Assert result.
+    report = getAssertionReport(result, expectation, {isDebug});
```

Comment:
i realize we say both 'expectation' and 'assertion' for the same thing. 
i prefer assertion but a little annoying to change that now. 

ah well.